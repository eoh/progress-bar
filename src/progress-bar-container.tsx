import * as React from 'react';
import ProgressBar from './progress-bar';

export interface ProgressBarContainerProps {
  /**
   * Defines the minimum value
   */
  min?: number;
  /**
   * Defines the maximum value
   */
  max?: number;
  /**
   * Defines the minimum value
   */
  onIncrement?: () => void;
  onDecrement?: () => void;
  /**
   * Defines the minimum value
   */
  progress?: number;
  /**
   * Determine if progress bar should self-increment
   */
  shouldSelfIncrement?: boolean;
  /**
   * Defines the interval value for incrementing
   */
  selfIncrementInterval?: number;
  /**
   * Defines the main progress bar color
   */
  fillColor?: string;
  /**
   * Defines the background color
   */
  bgColor?: string;
  /**
   * Defines the border style
   */
  borderStyle?: string;
  /**
   * Defines the border-radius
   */
  borderRadius?: number;
  /**
   * Defines the height
   */
  height?: number;
  /**
   * Defines the width
   */
  width?: number;
  /**
   * Defines the transition
   */
  transition?: string;
  /**
   * Renders increment and decrement buttons
   */
  demo?: boolean;
}

export interface ProgressBarContainerState {
  progress: number;
}

export default class ProgressBarContainer extends React.PureComponent<
  ProgressBarContainerProps,
  ProgressBarContainerState
  > {
    public intervalId: number;
    public timerId: number;

    constructor(props: ProgressBarContainerProps) {
    super(props);
    
    this.state = {
      progress: 0,
    };
  }

  render() {
    return (
      <div className="progress-bar-container">
        <ProgressBar 
          min={this.props.min}
          max={this.props.max}
          onIncrement={this.onIncrementClick}
          onDecrement={this.onDecrementClick}
          progress={this.state.progress}
          fillColor={this.props.fillColor}
          bgColor={this.props.bgColor}
          borderStyle={this.props.borderStyle}
          borderRadius={this.props.borderRadius}
          height={this.props.height}
          width={this.props.width}
          transition={this.props.transition}
          demo={this.props.demo}
        />
      </div>
    );
  }

  public componentDidMount() {
    if (this.props.shouldSelfIncrement) {
      this.selfIncrementInterval();
    }

    // Listen to external datasource
    // this.listenTo(datasource, 'externalEventName', this.handleExternalUpdate);
  }

  public componentWillUnmount() {
    if (this.intervalId) {
      clearInterval(this.intervalId);
    }
  }
  
  public componentDidUpdate(prevProps: ProgressBarContainerProps, nextState: ProgressBarContainerState) {
    return nextState.progress !== this.state.progress;
  }

  private selfIncrementInterval() {
    let percent = this.state.progress;
    this.intervalId = setInterval(this.selfIncrement, this.props.selfIncrementInterval);
    
    if (percent >= 100) {
      this.setState({
        progress: 99
      });
    } else {
      this.setState({
        progress: percent
      });
    }
  }

  private selfIncrement = () => {
    let percent = this.state.progress;
    percent += Math.random() + 5 - Math.random();
    percent = percent < 99 ? percent : 100;

    this.setState({
      progress: percent
    });
  }

  private onIncrementClick = () => {
    this.setState(nextState => ({
      progress: Math.min(Math.max(this.state.progress + 20, 1), 100)
    }));
  }

  private onDecrementClick = () => {
    this.setState(nextState => ({
      progress: Math.min(Math.max(this.state.progress - 20, 1), 100)
    }));
  }

  // 
  /* private handleExternalUpdate = (response) => {
    this.setState({
      progress: response
    });
  } */
}
