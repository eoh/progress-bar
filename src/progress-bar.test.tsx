import * as React from 'react';
import { configure } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

import { shallow, ShallowWrapper } from 'enzyme';
import ProgressBar from './progress-bar';
import { ProgressBarContainerProps } from './progress-bar-container';

describe('ProgressBar', () => {
  let component: ShallowWrapper<ProgressBarContainerProps, {}>;

  const getProps = (overrides?: Partial<ProgressBarContainerProps>) => {
    const defaultProps = {
      min: 20,
      max: 100,
      fillColor: '#aaa',
      bgColor: '#aaa',
      height: 50,
      width: 400,
      borderRadius: 20,
    };

    return {
      ...defaultProps,
      ...overrides,
    };
  };

  describe('when initialized', () => {
    let props: ProgressBarContainerProps;

    beforeEach(() => {
      props = getProps();
      component = shallow(<ProgressBar {...props} />);
    });

    it('matches snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
