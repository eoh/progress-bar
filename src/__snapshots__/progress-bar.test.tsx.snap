// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar when initialized matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProgressBar
    bgColor="#aaa"
    borderRadius={20}
    fillColor="#aaa"
    height={50}
    max={100}
    min={20}
    width={400}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="progress-bar--inner"
          style={
                    Object {
                              "backgroundColor": "#aaa",
                              "borderRadius": "20px",
                              "maxWidth": "100%",
                              "minWidth": "20%",
                              "transition": "undefined",
                              "width": "undefined%",
                            }
          }
/>,
        <span>
          20
           
</span>,
        <span>
          / 
          100
</span>,
        null,
      ],
      "className": "progress-bar",
      "data-progress": "10",
      "style": Object {
        "backgroundColor": "#aaa%",
        "border": "undefined%",
        "borderRadius": "20px",
        "height": "50px",
        "width": "400%",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "progress-bar--inner",
          "style": Object {
            "backgroundColor": "#aaa",
            "borderRadius": "20px",
            "maxWidth": "100%",
            "minWidth": "20%",
            "transition": "undefined",
            "width": "undefined%",
          },
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            20,
            " ",
          ],
        },
        "ref": null,
        "rendered": Array [
          20,
          " ",
        ],
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "/ ",
            100,
          ],
        },
        "ref": null,
        "rendered": Array [
          "/ ",
          100,
        ],
        "type": "span",
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="progress-bar--inner"
            style={
                        Object {
                                    "backgroundColor": "#aaa",
                                    "borderRadius": "20px",
                                    "maxWidth": "100%",
                                    "minWidth": "20%",
                                    "transition": "undefined",
                                    "width": "undefined%",
                                  }
            }
/>,
          <span>
            20
             
</span>,
          <span>
            / 
            100
</span>,
          null,
        ],
        "className": "progress-bar",
        "data-progress": "10",
        "style": Object {
          "backgroundColor": "#aaa%",
          "border": "undefined%",
          "borderRadius": "20px",
          "height": "50px",
          "width": "400%",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "progress-bar--inner",
            "style": Object {
              "backgroundColor": "#aaa",
              "borderRadius": "20px",
              "maxWidth": "100%",
              "minWidth": "20%",
              "transition": "undefined",
              "width": "undefined%",
            },
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              20,
              " ",
            ],
          },
          "ref": null,
          "rendered": Array [
            20,
            " ",
          ],
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "/ ",
              100,
            ],
          },
          "ref": null,
          "rendered": Array [
            "/ ",
            100,
          ],
          "type": "span",
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
